### ISSUE TITLE
Add Cancellation Support

### ISSUE THREAD

**Issue by thargy** [amitfin/retry#82]
### Checklist

- [X] I have filled out the template to the best of my ability.
- [X] This only contains 1 feature request (if you have multiple feature requests, open one feature request for each feature request).
- [X] This issue is not a duplicate feature request of [previous feature requests](https://github.com/amitfin/retry/issues?q=is%3Aissue+label%3A%22Feature+Request%22+).

### Is your feature request related to a problem? Please describe.

It would be nice to support cancellation. For example, when trying to turn off a light with a slow transition (e.g. 5s), if you then send a turn-on to the same entity with a fast transition (e.g. 1s), then when the validation for the first retry is checked after 5s, it will see it as 'failed' (the light was turned off by the 2nd command) and will turn the light back on again.

### Describe the solution you'd like

There are several approaches. The most flexible is to have a templatable string data field, e.g., `cancellation_tag`. When a retry command is received with this field set then any ongoing retries with the same tag will be cancelled.

You would then set it like so:

```yaml

service: retry.call
data:
  service: homeassistant.turn_off
  expected_state: "off"
  state_grace: 5.1
  transition: 5
  cancellation_tag: "light.kitchen"
target:
  entity_id: light.kitchen

# Sending on within 5s will cancel validation and retries of the first command

service: retry.call
data:
  service: homeassistant.turn_on
  expected_state: "on"
  state_grace: 1.1
  transition: 1
  cancellation_tag: "light.kitchen"
target:
  entity_id: light.kitchen
```

### Describe alternatives you've considered

You can achieve a similar effect using Script Modes, but this requires every set of commands to be embedded in a single script, which is significantly less useful.

### Additional context

This could be extended with a `cancel_with` tag, which would cancel retries if a template is evaluated as true.

However, that would probably be more flexibly useful as an `execute_when` tag, evaluated before the first execution and each subsequent retry. 

Both of these are useful to ensure you only continue trying whilst the initial condition that led to the execution remains true.

**NB** I haven't explained what `cancellation` means in detail, as I've not had time to evaluate your implementation fully.

At the fullest level, cancel would attempt to abort a command currently `executing`, though I'm not sure how feasible that is. If possible, that could make `cancel_with` and `execute_when` potentially subtly different - in that the `execute_when` would not cancel an ongoing attempt.

More likely, cancellation will just prevent validation and retry attempts, in which case `execute_when` is sufficient, as `cancel_with` is effectively just a `do...while`, where `execute_when` is a `while`.

**Comment by amitfin** [comment]
I'm thinking on the following approach:
- Allow only one retry call for a particular entity ID, i.e. a retry call automatically cancels a previous call on the same entity.
- Note that a retry loop is always running on a single entity, even when the original call has multiple entities (we create a separate loop per entity to isolate failures). This means that it is possible to use the entity ID as the cancelation key.
- The main benefit of this approach is that it doesn't require configuration.
- There might be edge cases where it makes sense to run more than one parallel loop on the same entity, but I'm unable to think on such a scenario. If it is common, we can add a cancelation key parameter which will override the default of using the entity ID as the cancelation key.

@thargy , WDTY?

**Comment by thargy** [comment]
> @thargy , WDTY?

Sorry for the delayed response; I wanted to give it considerable thought and present you with a solid, worked-through proposal. It might seem quite long-winded, but I've tried to be as clear and intuitive as possible.


I did consider that approach first; I leant towards the explicit key for the following reasons:

* Cancelling by default is a significant _breaking_ change—it will affect the action of existing code. In my experience of working with developers, asynchronous coding (including cancellation!) is notoriously hard for many developers to get right—never mind prosumers or gifted amateurs. Adding a breaking change would be hard for people to discover and debug.
* I went with a new tag, `cancellation_tag,` because it isn't a breaking change and includes the ability to cancel by entity ID—by convention, you can use the ID as the key. I considered the problem of configuration clash; hence, I suggested a long key name.
* Your approach doesn't allow cancelling the group of entities as a whole. For example, if you change 2 lights and then change 1 of the lights, only the task against the colliding light will cancel.

As you say, it might not be a bad thing to consider entities separately and not worry about grouping or being able to attribute actions to two separate entities. Indeed, my use case would be fine with your suggestion.

###  Counter proposal

You could argue that a better 'HA-like' route would be to add a `mode` attribute, like scripts and automations (see [Script Modes](https://www.home-assistant.io/integrations/script/#script-modes)). This would make it more accessible and in line with automation and scripts. That is a big win in terms of making it user-friendly.

Defaulting to `parallel` would be a non-breaking change as it would mimic what happens now. As you say, the current behaviour is ironically most likely to cause confusion (and result in the scenario I described). Having the ability to set the `parallel` option, though, would allow existing use cases to be 'fixed' back to the original behaviour.

Defaulting to `restart` would be the same as your proposal and probably give the most expected/useful behaviour, but it would be a breaking change.

Defaulting to `single` would be consistent with the actual script and automation modes, which both default to single. This could potentially be what pro-users would expect the default to be, but it would be a breaking change.

The problem with `mode` is that it gets complicated when two different calls point to the same entity id, as there is nothing to stop them from specifying a different `mode`:

```yaml
# Turn on the light in single mode
service: retry.call
data:
  service: homeassistant.turn_on
  mode: single
  expected_state: "on"
  state_grace: 1.1
  transition: 1
target:
  entity_id: light.kitchen

# No mode, so defaults to `restart`
service: retry.call
data:
  service: homeassistant.turn_off
  expected_state: "off"
  state_grace: 5.1
  transition: 5
target:
  entity_id: light.kitchen
```

This is not the case in automations and scripts, which have unique identities.

You need to decide what happens when the mode of the new task is different from that of an existing task. In the example above, the new task is in `restart` `mode`, but there is an existing task in `single` `mode`. The easiest solution is to error, but that is unhelpful. Further, with queueing, there may be multiple 'existing' tasks, one running, and many others in a queue.

A more intuitive approach is to modify the mode definitions subtly to reflect the ability to have existing tasks of different modes:
* `single` mode tasks will not run when an existing task is running and/or tasks are queued. Unlike automations and scripts, they can be interrupted by a `restart` task, but they will always issue a warning if they don't run, or are interrupted.
* `restart` mode tasks will always stop `single` or `queued` tasks, and start immediately. (Note `single` mode tasks will log a warning though whilst being interrupted)
* `queued` mode tasks will run immediately, or after all existing tasks are complete.
* `parallel` mode tasks run asynchronously and do not impact existing tasks.

 so a potential algorithm would look something like the following and should give reasonable behaviour:

```python
if mode = 'parallel':
	# Parallel tasks are a special case we just fork them, they never interact with existing tasks.
	<start task on new thread>
	return

# Check if there are any existing tasks.
if existing_tasks.count > 0:
	# Set the current highest mode to the new task's mode
	let highest_mode = mode
	
	# Loop through existing tasks, calculating the highest mode
	for task in existing_tasks
		# Update highest_mode based on 'queued' > 'restart' > 'single'
		# Note: We never put 'parallel' tasks in existing_tasks, so the
		# highest_mode will never be 'parallel'.
		if task.mode > highest_mode:
			highest_mode = task.mode
	
	# Now, decide what to do with this new task
	if highest_mode == 'single':
		# There should be only one existing task, and it's obviously in 'single' mode
		<log warning>
		# Do not continue running the new task as both are in 'single' mode
		# so the existing task takes priority
		return
	
	elif highest_mode == 'restart':
		if existing_tasks.count > 0:
			# We can only have one existing task, which must be in either 'single' or 'restart' mode.
			if existing_task[0].mode == 'single':
				# We should explain that a 'restart' task interrupted our 'single' mode task.
				<log warning>
	
			# Note that if both tasks were in 'restart' mode this new task will silently
			# interrupt the existing task as expected.
			existing_tasks.remove(task)
			<stop task>
		# Continue to run this task
	
	else:
		# We must have queued tasks at this point
		# We might have many existing tasks, only one will be either 'single' or 'restart' at this point.
		# What we do depends on our own mode
		if mode == 'single':
			# We shouldn't run our new task as there are existing tasks running
			# As this is the equivalent of a 'second call' indicate we did not run as there
			# were existing tasks
			<log warning>
			return
		elif mode == 'restart':
			# We should stop all existing tasks
			for task in existing_tasks:
				if task.mode == 'single':
					# We should explain that a 'restart' task interrupted our 'single' mode task.
					<log warning>

				# We stop 'restart' or 'queued' tasks silently
				existing_tasks.remove(task)
				<stop task>

			# Continue to run this task
		else:
			# We must be a 'queued' task, so add to the existing_tasks queue
			existing_tasks.add(this)
			# As we have existing tasks - we don't run yet.
			exit

# Add this task to our existing tasks
existing_tasks.add(this)

<run this task on this thread>

# We're done so remove the task
existing_tasks.remove(this)

if (existing_tasks.count > 0):
	# These should be queued tasks at this point
	<run next existing task>
```

With the above algorithm we can define an _optional_ `group_id`. In such a case it would override the default behaviour of grouping by _each_ target entity, and instead would group by every call with that group_id. For example,

```
# Turn on the lights in single mode
service: retry.call
data:
  group_id: dining_lights
  service: homeassistant.turn_on
  mode: single
  expected_state: "on"
  state_grace: 1.1
  transition: 1
target:
  entity_id:
	- light.kitchen
	- light.dining_room 


# No mode, so defaults to `restart`
service: retry.call
data:
  group_id: dining_lights
  service: homeassistant.turn_off
  expected_state: "off"
  state_grace: 5.1
  transition: 5
target:
  entity_id:
	- light.kitchen 
```

Here, the two lights would turn on, so long as there was no other existing tasks with the `dining_lights` `group_id` currently executing. Then, both tasks would be stopped (with a warning being logged), and the kitchen light would be switched off.

Another effect is that any other invocations affecting either light would be unaffected if their `group_id` was not set, or was set, or was different.

This addition would be relatively trivial to add if existing_tasks were stored by `group` and the `group_id` were effectively defaulted to each individual entity `unique_id`. I hope you can see it's potential usefulness.

**Comment by amitfin** [comment]
Thanks @thargy for the detailed thoughts. This is extremely helpful!

Version [2.9](https://github.com/amitfin/retry/releases/tag/v2.9.0) adds `retry_id` with defaults:
![image](https://github.com/amitfin/retry/assets/19599059/7f4d4bc5-f0ad-451f-930e-35f60d4a3793)

Note that this version is still in beta, so "Show beta versions" should be used in [HACS for downloading it](https://hacs.xyz/docs/faq/beta/).

@thargy , can you test it to check if it solves your scenario?

**Comment by thargy** [comment]
Hi, I will test over this weekend.

Sorry for another wall of text; you can probably see my background as a technical architect!

To ensure I've understood, I believe what you've done is the equivalent of implementing `group_id` (`retry_id` is a better name!) and `mode`=`restart` from my proposal above.

If so, I would slightly tweak your documentation on release to indicate the subtle difference between the default value of `retry_id` and an _explicit_ value. Specifically, where multiple entity_ids are targetted, then each entity will represent a separate `retry_id` instance; in contrast, an _explicit_ `retry_id` will mean that all such entity_id instances will be 'grouped' by a common retry_id. This won't be immediately obvious unless people understand that listing multiple targets will result in multiple independent retry calls, one for each entity (one entity might result in one service call, and a second entity might have multiple calls on failure). This ambiguity led me to refer to it as a `group_id`, but that was masking the primary purpose.

The result is that a retry call on a completely different entity _can_ cause an unrelated entity's retry call to be cancelled, _when using an explicit `retry_id`_. This a feature and use case for the explicit `retry_id`. Further, it ensures that all targets of a single call are cancelled en masse rather than individually. Finally, it can be used to _prevent_ a call from being cancelled by a different call targeting the same entity(ies).

Secondly, you should state whether setting `retry_id` to the entity_id (when there is only one entity_id) or service name _explicitly_ is equivalent to leaving the `retry_id` to be set _implicitly_. If so, then it should be made clear that setting the `retry_id` to an existing entity_id for calls which target other entities may have unintended effects (though potentially useful ones), for example:

```
# retry_id defaults to `light.kitchen`
service: retry.call
data:
  service: homeassistant.turn_on
  mode: single
  expected_state: "on"
  state_grace: 1.1
  transition: 1
target:
  entity_id:  light.kitchen


# Explicit retry_id means that the above call is cancelled by this one (and vice versa).
service: retry.call
data:
  retry_id: light.kitchen
  service: homeassistant.turn_off
  expected_state: "off"
  state_grace: 5.1
  transition: 5
target:
  entity_id: light.dining_room
```

The alternative, which has pros & cons, is that the default value cannot be _explicitly_ set. i.e. specifying a `retry_id` that matches an existing `entity_id` or service name will not cancel an implicitly set `retry_id`, but will cancel an _explicitly_ set `retry_id` that matches. I feel this latter scenario is preferable and less prone to errors, and am happy to justify further, if you would find it helpful. (Implementation is pretty simple, you can efficiently disambiguate in a dictionary structure using an invalid character prefix, or similar tricks)

Overall, I think your choice is a legitimate way forward. However, you may consider partial implementation of `mode` to support both `restart` (as you've done) and `parallel`  (existing, non-cancellation logic). This wouldn't be too difficult and would allow any existing logic broken by this change to be restored by adding `mode: parallel`. Of course, that wouldn't be a breaking change from the above release if `mode: restart` is considered the default, and it would be extensible to support the other modes (as above) in future. It would also nudge users (who may not be Computer Science majors, and unfamiliar with asynchronous programming) to another area where they've seen examples of cancellation already.

**Comment by amitfin** [comment]
The main goal of the design is to "just work" for the majority (or all) of the use cases. It's not clear to me if there are actual use cases that require using `retry_id`, but I've kept it as an option just in case... (and for backward compatibility, if I'm missing something here).

You are correct that this can be seen as `mode: restart`. Note that it's possible to have the equivalent of `mode: parallel` by setting different `retry_id` parameters.

The documentation was updated based on your recommendations:
https://github.com/amitfin/retry?tab=readme-ov-file#retry_id-parameter-optional

![image](https://github.com/amitfin/retry/assets/19599059/194bed0a-01f3-4865-9b8a-0abc5da5c0ba)

Thoughts?

**Comment by thargy** [comment]
That's really well written! And probably all that's necessary.

To be clear, this implies that _explicitly_ setting the `retry_id` to an `entity_id` or 'service name' can collide with an _implicitly_ set default, is that right?

> The main goal of the design is to "just work" for the majority (or all) of the use cases. It's not clear to me if there are actual use cases that require using retry_id, but I've kept it as an option just in case... (and for backward compatibility, if I'm missing something here).

Fair enough. My design process is always to figure out the most feature-rich version I can and then create an MVP that doesn't preclude it if necessary. Your proposal is a very reasonable MVP and will probably be sufficient in most use cases, so it adheres well to the KISS philosophy.

> Note that it's possible to have the equivalent of mode: parallel by setting different retry_id parameters.

Yes, this would have mostly the same effect, though it relies on the user using a unique ID in each case, which can be difficult in this scenario as there's no central repository of retry calls - which are likely to be defined all over the place (automation, scripts, etc.). Importantly, it _doesn't_ prevent the retry call from cancelling _itself_ (unlike `parallel`) _unless_ you passed a different unique ID on _each_ invocation (which is probably beyond most users). A good practice would be to use a template that provides a GUID identifier to prevent a retry call from being cancelled. Still, the average user is unlikely to think of that or provide a robust implementation. 

However, as most users probably won't even notice the change nor wonder how to prevent cancellation, a good compromise is to add a suggestion to the notes section, like:

> Although the addition of cancellation support via `retry_id` was a breaking change in version 2.9, you can effectively prevent cancellation of a retry call, restoring previous functionality, by providing a GUID for each invocation, e.g.

```jinja
retry_id: >
  {% for i in range(0,32) -%}{{ [0,1,2,3,4,5,6,7,8,9,"a","b","c","d","e","f"]|random }}{% endfor %}
```

If there was a lot of demand for preventing cancellation, you could easily add this via the GUI for non-developers. Further, your current solution of adding the `retry_id` with notes does not prevent the future addition of a `mode` if a solid use case arises.

Thanks for enduring my ramblings!

**Comment by amitfin** [comment]
A new option to disable the cancellation logic was added (v2.9.2):

https://github.com/amitfin/retry?tab=readme-ov-file#retry_id-parameter-optional

<img width="870" alt="image" src="https://github.com/amitfin/retry/assets/19599059/17d7afd1-d992-4a8d-bfdd-1352e83403ad">

**Comment by thargy** [comment]
The only downside I can see to such an approach is that it would potentially be duplicate functionality if you were to add `mode` in future. That is a minor criticism.

However, it would be very useful in the absence of `mode`, so it would be preferable for this MVP.

I would emphasise it's utility in 'restoring the existing functionality'.

**Comment by amitfin** [comment]
v2.9.x is running for a few days now, so making it an official release (not in beta mode anymore).

**Comment by amitfin** [comment]
I found a severe issue with the original implementation: multiple retry loops of the same service call (either multiple entities or multiple service calls via retry.actions) would cancel each other when having the same retry_id (when it's set explicitly). This was fixed by v2.9.3. The solution was to have the same loop ID for all loops which were generated by the same service call (retry.call or retry.actions). In addition, the mechanism now supports reference counting, meaning retry ID can have multiple (identical) loop IDs.
